airflow/providers/amazon/aws/hooks/base_aws.py

```diff
@@ -62,16 +63,14 @@ class AwsBaseHook(BaseHook):
 
     def __init__(
             self,
-            aws_conn_id="aws_default",
+            aws_conn_id: Optional[str] = "aws_default",
```

Comment:
I think we should not use `Optional ` here. Should only use in default `None`, am I wrong?

Reply:
Optional is a valid type annotation here because we allow user to explicitly pass None as the value.

Reply:
I not sure about that, I init think it should be use only if default value equal to `None`

Reply:
@turbaszek @potiuk https://docs.python.org/3/library/typing.html#typing.Optional WDYT

Reply:
@zhongjiajie is right:
> Optional[X] is equivalent to Union[X, None].

we should use `aws_conn_id: str = "aws_default"` because passing `None` will cause an error (probably)

Reply:
This does not cause error for AWS.  AWS hooks should support falling back to using default credential chain lookup behavior when connection id is not specified.

Reply:
> This does not cause error for AWS. AWS hooks should support falling back to using default credential chain lookup behavior when connection id is not specified.

That's why I used "probably" because I don't know AWS hook behavior. Still we should use this:
```
aws_conn_id: str = "aws_default"
```
or this:
```
aws_conn_id: Optional[str] = None
```

Reply:
> This does not cause error for AWS. AWS hooks should support falling back to using default credential chain lookup behavior when connection id is not specified.

!!!!!!! Wow, I don't know aws have this behavior!

Reply:
This PR is just about to bring back this behavior. `aws hook should work without conn id`

Reply:
For GCP, we used a different behavior GCP always need a connection because it defines connection information. A special connection mode is ADC expressed by connection configuration. ADC is the default configuration in Airflow.
```
    `Application Default Credentials`_ provides an easy way to obtain
    credentials to call Google APIs for server-to-server or local applications.
    This function acquires credentials from the environment in the following
    order:

    1. If the environment variable ``GOOGLE_APPLICATION_CREDENTIALS`` is set
       to the path of a valid service account JSON private key file, then it is
       loaded and returned. The project ID returned is the project ID defined
       in the service account file if available (some older files do not
       contain project ID information).
    2. If the `Google Cloud SDK`_ is installed and has application default
       credentials set they are loaded and returned.

       To enable application default credentials with the Cloud SDK run::

            gcloud auth application-default login

       If the Cloud SDK has an active project, the project ID is returned. The
       active project can be set using::

            gcloud config set project

    3. If the application is running in the `App Engine standard environment`_
       then the credentials and project ID from the `App Identity Service`_
       are used.
    4. If the application is running in `Compute Engine`_ or the
       `App Engine flexible environment`_ then the credentials and project ID
       are obtained from the `Metadata Service`_.
    5. If no credentials are found,
       :class:`~google.auth.exceptions.DefaultCredentialsError` will be raised.

    .. _Application Default Credentials: https://developers.google.com\
            /identity/protocols/application-default-credentials
    .. _Google Cloud SDK: https://cloud.google.com/sdk
    .. _App Engine standard environment: https://cloud.google.com/appengine
    .. _App Identity Service: https://cloud.google.com/appengine/docs/python\
            /appidentity/
    .. _Compute Engine: https://cloud.google.com/compute
    .. _App Engine flexible environment: https://cloud.google.com\
            /appengine/flexible
    .. _Metadata Service: https://cloud.google.com/compute/docs\
            /storing-retrieving-metadata
```

Reply:
There is only one exception to this rule in Airflow 1.10.x, which has been fixed in Airflow 2.0 - GKEPodOperator.

Reply:
Authorization using the metaserver is particularly problematic, because the server may have wider rights than we expect the instance user to have. Example: The server may have read / write access to Stackdriver logs, but the user should not have these privileges. The similar thing is with system buckets that the server can use, but the user should not be able to modify them. A similar thing looks like system buckets that the server can use, but the user should not be able to modify them.

Reply:
the wording in the documentation linked by @zhongjiajie is a little bit misleading, which probably caused confusion between `Optional` type and optional argument.

`def foo(a: Optional[str] = 'test'):` tells mypy that argument `a` needs to be either `None` or a `str`. Since the default value `'test` is a string, this is a valid use of `Optional` type annotation. `a: Optional[str] = None` is also a valid annotation.

If we allow `a` to take `None` as value, but annotate it as `def foo(a: str='test')`, then mypy will not be able to catch errors like `b = a + '_suffix'` because the annotation `str` tells mypy that `a` will always be a string. Mypy will also prevent users from calling `foo` with `foo(None)` due to missing `Optional` in the type annotation, which is not what we want.

Behavior wise, both `aws_conn_id: Optional[str] = None` and `aws_conn_id: Optional[str] = 'aws_default'` will work for aws hook. I kept it to `aws_conn_id: Optional[str] = 'aws_default'` because that's what it used to be. I don't know if it's going to break anything if default is changed to `None` since there might be users out there actually using `aws_default` connection as the way to inject global aws credential instead of using default credential chain lookup.

In AWS land, the best practice is to use default credential chain lookup, so `aws_conn_id: Optional[str] = None` would be better. It's up to us whether we are willing risk breaking existing users in order to adopt a better practice. I don't have a strong opinion on this, but if anyone thinks we should take the chance for airflow 2.0, I am open to change it to `None`.

Reply:
> the wording in the documentation linked by @zhongjiajie is a little bit misleading, which probably caused confusion between `Optional` type and optional argument.

I can quite agree. In fact, I think introducing types in dynamically typed langauge creates more confusion that in statically typed one. In most places in airflow, I think we are using `Optional` when the default value is `None`

Reply:
I also do not have a strong opinion, but I wanted to present the GCP approach for comparison.

Reply:
> > the wording in the documentation linked by @zhongjiajie is a little bit misleading, which probably caused confusion between `Optional` type and optional argument.
> 
> I can quite agree. In fact, I think introducing types in dynamically typed langauge creates more confusion that in statically typed one. In most places in airflow, I think we are using `Optional` when the default value is `None`

Yep, It's a little confusion here, thanks for the clarification. And maybe we better keep default value not equal to None.